contracts/modules/cover/CoverOrder.sol

```diff
@@ -0,0 +1,225 @@
+import "@openzeppelin/contracts-v4/access/Ownable.sol";
+
+import "../../interfaces/ICoverOrder.sol";
+import "../../interfaces/IMemberRoles.sol";
+import "../../interfaces/INXMMaster.sol";
+import "../../interfaces/INXMToken.sol";
+import "../../interfaces/IWeth.sol";
+import "../../interfaces/IPool.sol";
+
+contract CoverOrder is ICoverOrder, Ownable {
+  using ECDSA for bytes32;
+  using SafeERC20 for IERC20;
+
+  /* ========== STATE VARIABLES ========== */
+  mapping(bytes32 => bool) public executedOrders;
+
+  /* ========== IMMUTABLES ========== */
+  ICover public immutable cover;
+  IMemberRoles public immutable memberRoles;
+  INXMToken public immutable nxmToken;
+  INXMMaster public immutable master;
+  IWeth public immutable weth;
+
+  /* ========== CONSTANTS ========== */
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  uint private constant ETH_ASSET_ID = 0;
+  uint private constant NXM_ASSET_ID = type(uint8).max;
+
+  /* ========== CONSTRUCTOR ========== */
+  constructor(
+    address _cover,
+    address _memberRoles,
+    address _nxmToken,
+    address _master,
+    address _weth,
+    address _owner
+  ) {
+    master = INXMMaster(_master);
+    cover = ICover(_cover);
+    memberRoles = IMemberRoles(_memberRoles);
+    nxmToken = INXMToken(_nxmToken);
+    weth = IWeth(_weth);
+    transferOwnership(_owner);
+  }
+
+  /// @notice Verifies and executes the order to buy cover on behalf of the creator of limit order
+  /// @param params Cover buy parameters
+  /// @param poolAllocationRequests Pool allocations for the cover
+  /// @param limitOrderId The ID of the limit order
+  /// @param signature The signature of the order
+  function executeOrder(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests,
+    uint limitOrderId,
+    bytes calldata signature
+  ) external payable returns (uint coverId) {
+
+    if (params.owner == address(0) || params.owner == address(this)) {
+      revert InvalidOwnerAddress();
+    }
+
+    if (params.paymentAsset == NXM_ASSET_ID) {
+      revert InvalidPaymentAsset();
+    }
+
+    // Compute hash
+    bytes32 orderHash = keccak256(
+      abi.encode(
+        params.productId,
+        params.coverId,
+        params.amount,
+        params.period,
+        params.paymentAsset,
+        params.coverAsset,
+        params.owner,
+        limitOrderId
+      )
+    );
+
+    // Ensure the order has not already been executed
+    if (executedOrders[orderHash]) {
+      revert OrderAlreadyExecuted(limitOrderId);
+    }
+
+    // Recover the signer address from the signature
+    bytes32 ethSignedOrderHash = orderHash.toEthSignedMessageHash();
+    address signer = ethSignedOrderHash.recover(signature);
+
+    // Ensure the signer is the user who owns the order
+    if (signer != params.owner) {
+      revert InvalidSignature();
+    }
+
+    // Mark the order as executed
+    executedOrders[orderHash] = true;
```

Comment:
It is actually recommended to mark the signature as used instead of something else to avoid signature malleability issues (can't see a specific issue here but it's better to be extra cautious).